### PR TITLE
Remove `mem_info` and references to `parity-util-mem`

### PR DESCRIPTION
**Pull request by m-cat**
# PULL REQUEST

## Overview

See title. This PR removes `parity-util-mem`. It upgrades some dependencies to fully remove it from `Cargo.lock`. 

## Companion

polkadot companion: https://github.com/paritytech/polkadot/pull/6374

## Related Issues

https://github.com/paritytech/substrate/issues/12658

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2094966

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2094965

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2094960

File: scripts/ci/deny.toml [hunk 1]
```diff
@@ -160,7 +160,7 @@ allow = [
 ]
 # List of crates to deny
 deny = [
-    { name = "parity-util-mem", version = "<0.6" }
+    { name = "parity-util-mem" }
```
Comment (ordian) [inline scripts/ci/deny.toml:163]: just remove it since we're removing the crate completely
```suggestion
```

File: bin/node/bench/Cargo.toml [hunk 2]
```diff
@@ -23,8 +23,8 @@ sp-state-machine = { version = "0.13.0", path = "../../../primitives/state-machi
 serde = "1.0.136"
 serde_json = "1.0.85"
 derive_more = { version = "0.99.17", default-features = false, features = ["display"] }
-kvdb = "0.12.0"
-kvdb-rocksdb = "0.16.0"
+kvdb = { version = "0.12.0", path = "../../../../parity-common/kvdb" }
+kvdb-rocksdb = { version = "0.16.0", path = "../../../../parity-common/kvdb-rocksdb" }
```
Comment (ordian) [inline bin/node/bench/Cargo.toml:27]: please revert, we can update it in a separate PR

**Comment by ordian** [comment]
> Depends on https://github.com/paritytech/parity-common/pull/696 being merged, first.

I'd suggest to remove the references first, then update and publish kvdb&co and then update it in substrate.

**Comment by m-cat** [comment]
@ordian If we just remove the references without pointing to the updated `kvdb`, we will get compile errors. See https://github.com/paritytech/substrate/issues/12658#issuecomment-1327870283. It's just a benchmark that fails to build, but I think it would break CI.

**Comment by ordian** [comment]
Fair enough, let's do it starting with polkadot's runtime-api cache, then parity-common, trie, substrate + polkadot + cumulus.

**Comment by ordian** [comment]
New versions of parity-common and trie crates have been published.
`cargo upgrade --workspace  memory-db kvdb kvdb-rocksdb kvdb-memorydb trie-bench` should do the trick (required [cargo-edit](https://github.com/killercup/cargo-edit)).

**Comment by ordian** [comment]
As mentioned in https://github.com/paritytech/substrate/issues/12658#issuecomment-1309746085, in polkadot (and cumulus?) companions, you'd need to set `jemalloc` as global allocator ([example](https://github.com/paritytech/parity-common/blob/806ca48a95c06014e0fde74a1382dc128da62206/parity-util-mem/src/lib.rs#L25-L26)).

**Comment by m-cat** [comment]
> New versions of parity-common and trie crates have been published. `cargo upgrade --workspace memory-db kvdb kvdb-rocksdb kvdb-memorydb trie-bench` should do the trick (required [cargo-edit](https://github.com/killercup/cargo-edit)).

I needed to modify this command (latest `cargo-edit`):

```
cargo upgrade -p memory-db -p kvdb -p kvdb-rocksdb -p kvdb-memorydb -p trie-bench --incompatible
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2099707

File: Cargo.lock [hunk 3]
```diff
@@ -68,16 +68,16 @@ version = "0.7.6"
...
- "getrandom 0.2.3",
+ "getrandom 0.2.8",
  "once_cell",
  "version_check",
 ]
 
 [[package]]
 name = "aho-corasick"
-version = "0.7.18"
+version = "0.7.20"
```
Comment (ordian) [inline Cargo.lock:78]: We prefer not to have `cargo update` and review individual dependabot updates.

File: client/informant/src/lib.rs [hunk 4]
```diff
@@ -56,13 +55,13 @@ impl Default for OutputFormat {
 pub async fn build<B: BlockT, C, N, P>(
 	client: Arc<C>,
 	network: N,
-	pool: Arc<P>,
+	_pool: Arc<P>,
```
Comment (ordian) [inline client/informant/src/lib.rs:58]: if it's unused now, should we remove this arg?

File: scripts/ci/deny.toml (hunk 1 above)
Comment (ordian) [reply scripts/ci/deny.toml:163]: still holds ^^^

**Review by ordian — COMMENTED** [review]
Does not compile yet.

File: Cargo.lock (hunk 3 above)
Comment (m-cat) [reply Cargo.lock:78]: Ah oops, the `cargo upgrade` command wasn't working so I tried `cargo update`. I always get those mixed up, quite confusing IMO.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2099817

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2099908

**Comment by m-cat** [comment]
> As mentioned in [#12658 (comment)](https://github.com/paritytech/substrate/issues/12658#issuecomment-1309746085), in polkadot (and cumulus?) companions, you'd need to set `jemalloc` as global allocator ([example](https://github.com/paritytech/parity-common/blob/806ca48a95c06014e0fde74a1382dc128da62206/parity-util-mem/src/lib.rs#L25-L26)).

n00b question, where should this be set? `src/main.rs`?

**Comment by ordian** [comment]
> n00b question, where should this be set? `src/main.rs`?

AFAIK, not necessarily. The only requirement is that it's set once in a crate or its recursive dependencies.
Not entirely sure what's the best place for it, but main will do.

Another thing is https://github.com/paritytech/polkadot/blob/71a0fadd627e97b5a71c3cc6fa2aefda14328c19/node/overseer/src/lib.rs#L657. We could preserve this by copying code from https://github.com/paritytech/parity-common/blob/fc75eaad1bb6f1cc40d6b7aa6a3b9093a841eddc/parity-util-mem/src/memory_stats_jemalloc.rs.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2101254

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2101911

**Comment by m-cat** [comment]
Error in CI:

```
check_dependent_project
========================
This check ensures that this project's dependents do not suffer downstream breakages from new code
[...]
error: failed to select a version for `sp-trie`.
    ... required by package `polkadot-cli v0.9.33 (/builds/parity/mirrors/substrate/polkadot/cli)`
versions that meet the requirements `*` (locked to 7.0.0) are: 7.0.0
the package `polkadot-cli` depends on `sp-trie`, with features: `memory-tracker` but `sp-trie` does not have these features.
failed to select a version for `sp-trie` which could resolve this conflict
```

Not sure what is usually done to resolve this.

File: bin/node/cli/Cargo.toml [hunk 5]
```diff
@@ -101,9 +101,7 @@ try-runtime-cli = { version = "0.10.0-dev", optional = true, path = "../../../ut
 serde_json = "1.0.85"
 
 [target.'cfg(any(target_arch="x86_64", target_arch="aarch64"))'.dependencies]
-sp-trie = { version = "7.0.0", default-features = false, path = "../../../primitives/trie", features = [
-	"memory-tracker",
-] }
+sp-trie = { version = "7.0.0", default-features = false, path = "../../../primitives/trie" }
```
Comment (bkchr) [inline bin/node/cli/Cargo.toml:104]: The dependency isn't used as far as I can see. I think it was only here to enable the feature.

**Review by bkchr — APPROVED** [review]

File: bin/node/cli/Cargo.toml (hunk 5 above)
Comment (m-cat) [reply bin/node/cli/Cargo.toml]: Seems like it.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2102551

**Review by ordian — APPROVED** [review]
Awesome! :tada:

**Review by sergejparity — APPROVED** [review]

**Review by altaua — APPROVED** [review]

**Comment by ordian** [comment]
bot notice me senpai

**Comment by mordamax** [comment]
bot merge

**Comment by ordian** [comment]
> Github API says polkadot#6374 is not mergeable

@m-cat I suspect merging master into companion might help

**Comment by m-cat** [comment]
> > Github API says polkadot#6374 is not mergeable
> 
> @m-cat I suspect merging master into companion might help

I've been trying and re-running the jobs haha, seems like something's always broken. Looks like the latest merge didn't help. 😛 Not sure if this is normal or just bad luck.

**Comment by ordian** [comment]
bot merge

**Comment by ordian** [comment]
bot merge

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis-v0-9-36/1529/1

**Pull request merged**